Sources/NIOCore/AsyncSequences/NIOHighLowWatermarkBackPressuredAsyncSequenceStrategy.swift

```diff
@@ -0,0 +1,45 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+/// A high-low watermarked back-pressure strategy for a ``NIOAsyncSequenceProducer``.
+///
+/// This strategy does the following:
+/// - On yield it keeps on demanding more as long as the `highWatermark` isn't reached.
+/// - On next it starts to demand once the the `lowWatermark` is reached.
+public struct NIOHighLowWatermarkAsyncSequenceProducerBackPressureStrategy: NIOAsyncSequenceProducerBackPressureStrategy {
```

Comment:
This is quite a name...  

How about nesting it inside `NIOAsyncSequenceProducer` as `WatermarkedStrategy`?

Reply:
I thought about this but decided against it for now since I had the feeling that it is very undiscoverable and it doesn't make it much better since in the end you have to spell out `NIOAsyncSequenceProducer<Element, NIOAsyncSequenceProducer.WatermarkedStrategy, Delegate>` open to change this though!

Reply:
I don't hate the name... it's not going to be spelled much, but the nested version does read nicer! :-)

I'd propose `NIOAsyncSequenceProducer.HighLowWatermarkStrategy` - the "high low" bit really explains what this is; I would not be trimming it out of the name; though the nesting helps make it a bit easier to read, I'd +1 that